### PR TITLE
[web] Fix window.defaultRouteName

### PR DESCRIPTION
**Pull request by mdebbar**
This PR contains 2 fixes:
1. Sticky `window.defaultRouteName` was causing routing issues after navigating in the app.
1. Stop prepending "/" to all route names (i.e. allow routes that don't start with a slash to work properly in Flutter).

Fixes https://github.com/flutter/flutter/issues/50639

This might also fix the problems seen in this issue: https://github.com/flutter/flutter/issues/43982 but it's hard to tell since there's no consistent repro.

### CODE REVIEW

File: lib/web_ui/lib/src/engine/browser_location.dart [hunk 1]
```diff
@@ -96,9 +75,7 @@ class HashLocationStrategy extends LocationStrategy {
     // Dart will complain if a call to substring is
     // executed with a position value that exceeds the
     // length of string.
-    path = path.isEmpty ? path : path.substring(1);
-    // The path, by convention, should always contain a leading '/'.
-    return ensureLeading(path, '/');
+    return path.isEmpty ? path : path.substring(1);
```
Comment (ferhatb) [inline lib/web_ui/lib/src/engine/browser_location.dart:78]: add assert(path[0] == '') before you substring ?

Comment (mdebbar) [reply lib/web_ui/lib/src/engine/browser_location.dart:78]: I'm not sure I understand. `path` could either be an empty string, or a string that starts with `#`. Did you mean `assert(path[0] == '#')`?

Comment (mdebbar) [reply lib/web_ui/lib/src/engine/browser_location.dart:79]: Ok, I added `assert(path.isEmpty || path.startsWith('#'));`

**Comment by mdebbar** [comment]
Re-ran the `Windows Host Engine` task in LUCI and it passed but Github didn't pick it up for some reason. Merging..

**Pull request merged**